### PR TITLE
http3: allow concurrent calls to Body.Close

### PR DESCRIPTION
**Pull request by RPRX**
Re-apply https://github.com/quic-go/quic-go/pull/4638

### CODE REVIEW

**Comment by marten-seemann** [comment]
What is "more graceful" about this? Please describe the bug this is supposed to fix (or even better: open an issue for that).

Please also add a test case.

**Comment by RPRX** [comment]
> What is "more graceful" about this? Please describe the bug this is supposed to fix (or even better: open an issue for that).

If the requestDone() is called concurrently, **the program may panic, because "close of closed channel"**: https://github.com/XTLS/Xray-core/issues/4148

File: http3/body.go [hunk 1]
```diff
@@ -96,7 +97,7 @@ type hijackableBody struct {
 	// The channel is closed when the user is done with this response:
 	// either when Read() errors, or when Close() is called.
 	reqDone       chan<- struct{}
-	reqDoneClosed bool
+	reqDoneClosed sync.Once
```
Comment (marten-seemann) [inline http3/body.go:100]:
```suggestion
	reqDoneOnce sync.Once
```

**Review by marten-seemann — COMMENTED** [review]
Please also add a test case.

**Comment by RPRX** [comment]
> Please also add a test case.

I'm not sure if we can make real concurrent calls that must call "close(r.reqDone)" concurrently.

Please feel free to edit the code / the commit message directly.

**Comment by RPRX** [comment]
And I'm not sure if `if r.reqDone == nil {` should be deleted.

**Comment by marten-seemann** [comment]
> I'm not sure if we can make real concurrent calls that must call "close(r.reqDone)" concurrently.

Sure, you can spawn 2 Go routines that both call `Close`. While that's not guaranteed to hit the race condition every time, at least from time to time it will.

**Comment by RPRX** [comment]
I've added a line of comment, so that developers will handle these code carefully. I think it is more useful than a random test.

**Comment by marten-seemann** [comment]
This comment doesn't add any value. Please remove it. And please add a test case, as I've requested for the third time now. It is useful, and this PR will not be merged without a test.

**Comment by RPRX** [comment]
Well, I've add a test case, though "that's not guaranteed to hit the race condition every time". Please review it.

**Review by marten-seemann — APPROVED** [review]

File: http3/body_test.go [hunk 2]
```diff
@@ -54,6 +54,15 @@ var _ = Describe("Response Body", func() {
 		Expect(rb.Close()).To(Succeed())
 	})
 
+	It("allows concurrent calls to Close", func() {
+		str := mockquic.NewMockStream(mockCtrl)
+		rb := newResponseBody(&stream{Stream: str}, -1, reqDone)
+		str.EXPECT().CancelRead(quic.StreamErrorCode(ErrCodeRequestCanceled)).MaxTimes(2)
+		go Expect(rb.Close()).To(Succeed())
```
Comment (marten-seemann) [inline http3/body_test.go:61]:
```suggestion
		go func() {
			defer GinkgoRecover()
			Expect(rb.Close()).To(Succeed())
		}()
```

[the remark this answers is not in the document]
Comment (marten-seemann) [reply http3/body_test.go:61]: @RPRX Can you apply this? For some reason, GitHub doesn't allow me to push to the branch.

**Comment by RPRX** [comment]
@marten-seemann It's ready.

**Review by marten-seemann — APPROVED** [review]
Thank you for your contribution @RPRX!

**Pull request merged**